source/d3d8to9_device.cpp

```diff
@@ -36,13 +36,14 @@ HRESULT STDMETHODCALLTYPE Direct3DDevice8::QueryInterface(REFIID riid, void **pp
 }
 ULONG STDMETHODCALLTYPE Direct3DDevice8::AddRef()
 {
-	_ref++;
+	InterlockedIncrement( &_ref );
 
 	return _proxy->AddRef();
 }
 ULONG STDMETHODCALLTYPE Direct3DDevice8::Release()
 {
-	if (--_ref <= 2)
+	ULONG myRef = InterlockedDecrement( &_ref );
+	if (myRef <= 2)
```

Comment:
Because the device holds two object instances in `_current_rendertarget` and `_current_depthstencil` which in turn each hold a reference to the device. Because of this the reference count should always be exactly two rather than zero after the application cleaned up all its resources.